### PR TITLE
fix(hybrid) more robust clustering connection and thread management

### PR DESCRIPTION
**Pull request by bungle**
### Summary

Implements a more robust websockets connection and thread management for clustering endpoint.

1. Tries to properly close the connection and notify the client
2. Tries to properly kill the started threads on reconnection
3. Adjusts so that expected events are not logged as errors
   - Timeouts
   - Closing of connection
4. Does not start zombie threads

Fixes an issue reported by other team.

This also contains additional commit to not cause push configs on clustering_data_planes events.

### CODE REVIEW

**Comment by dndx** [comment]
Hold off on merging for a moment, pending sync up with @bungle.

**Comment by kidd** [comment]
What's the status on this one?

**Comment by bungle** [comment]
@kidd, well... I guess you forked our test suite at some point, but then have not synched the changes. So this issue is partly because of not synching your suite. The main culprit is here:
https://github.com/Kong/kong-ee/blob/next/2.2.x.x/spec-ee/02-integration/14-hybrid_mode/01-synch_spec.lua#L51-L59

That file was copy pasted from:
https://github.com/Kong/kong/blob/master/spec/02-integration/09-hybrid_mode/01-sync_spec.lua

See what we added there:
https://github.com/Kong/kong/blob/next/spec/02-integration/09-hybrid_mode/01-sync_spec.lua#L24

So if you want quick solution, just add that while we figure out what to do with this PR.

**Comment by kidd** [comment]
Ah I see! I guess we created an ee-only test, but then if defaults changed we didn't notice it. 

Unfortunately, we don't have any way to sync at that level. 

We're going to add that today so we can keep merging in. Thanks for unblocking!

cc @mikefero

File: kong/clustering.lua
Comment (dndx) [inline kong/clustering.lua:121]:
```suggestion
    ngx_log(is_timeout(err) and ngx_NOTICE or ngx_WARN, "unable to ping control plane node: ", err)
    return
```

File: kong/clustering.lua
Comment (dndx) [inline kong/clustering.lua:462]:
```suggestion
  wb:close()
```

We should just close the WS connection here, if `send_close` hasn't been called, then the `close` method will call it automatically. Because the other end might be dead by now, and `send_close` may keep the (dead) connection around for longer than we need.

File: kong/clustering.lua
Comment (dndx) [inline kong/clustering.lua:229]: So, for the DP side of code, when `c:recv_frame` timedous, we should also be checking to see if this has been overdue compared to `PING_WAIT`, just like the CP did. These code can largely be duplicated from the CP side.

File: kong/clustering.lua
Comment (dndx) [inline kong/clustering.lua:412]: Maybe call this `send_thread` and the `ping_thread` -> `read_thread` instead, it may makes the code easier to follow for us.

**Review by dndx — COMMENTED** [review]
Left some comments, also I pushed out a docs commit 84560666bbb18003dc8f549cdb896881086f149b so that the behavior is better documented for future reference.

**Comment by dndx** [comment]
Might need to rebase onto `master` as well.

**Review by dndx — APPROVED** [review]
LGTM, let's make sure to confirm the status of RC1 release before merging.

**Pull request merged**
